README.md
# ng-roo
ng-roo is an offline storage strategy for angular apps, leveraging PouchDB
(see what we did there?) and CouchDB, and emphasizing local storage as the single
source of truth.

## Prerequisites
* PouchDB
* moment.js
* lodash.js

## Usage
To use ng-roo, simply include it in your app's list of dependencies.

```
angular.module('MyOfflineApp', ['vs.ng-roo'])
  .config(function(rooConfigProvider) {
	  // Our local database configuration will go here
  });
```

### Sinks and Sources
ng-roo operates by having read-only source databases and write-only sink
databases. To configure databases, we have to make rooConfigProvider aware of
our CouchDB instance url, and our source databases ("down databases") as well as
our sink databases ("up databases").

```
angular.module('MyOfflineApp', ['vs.ng-roo'])
  .config(function(rooConfigProvider) {
	  rooConfigProvider.couchConfig({ 'couchUrl': 'http://0.0.0.0:5984' });
	  rooConfigProvider.dbs(['myReadOnlyDb1', 'myReadOnlyDb2'], ['myWriteDb']);
  });
```

You will, of course, want to create CouchDB databases for each database you
pass to `rooConfigProvider.dbs`, and have your CouchDB instance running,
otherwise you will get errors. See the PouchDB documentation for information
on Pouch/Couch syncing.

To use the offline storage ng-roo provides, simply inject the `Pouch` object,
and create a new instance of a database, passing the name of the database as a
string. ng-roo will create the database, if not already created, or it will
pass the already-created database back. You can then call api methods on the
database. ng-roo uses IndexedDB under the covers.

```
angular.module('MyOfflineApp').service('MyService', function(Pouch) {
  function getDocs() {
		var myDb = new Pouch('myReadOnlyDb1');
		myDb.getAll().then(function(docs) {
		  doSomethingWithDocs(docs);
		});
	}
});
```

Under the covers, ng-roo syncs data outstanding in the write db(s) with the
retrieved docs in the read dbs to present a consistent data model to the user.
You would most likely use this in conjunction with your own sync service that
would call ng-roo's sync finished hooks to remove records from the write
database(s).

To write to an ng-roo sink/up database, you create an instance of the database
similar to above:

```
var myJournalDb = new Pouch('myWriteDb');
```

Next, call `putEntry()` passing the following parameters:
* readOnlyDbName (string)
* record id (string)
* change (an arbitrary json object consisting of whatever you want recorded)
* HTTP method (string)
* URL (string)
* Data (string) -- for an HTTP POST, the data in the request body
* Headers (string) -- any header information needed to be processed by the server
* User (json object) -- any user information needed for authorization/authentication.
* Success callback (function)
* Error callback (function)

ng-roo writes the record to the database. Note that whatever syncing process
you employ with your server will then have the information to take this record
and process it with its HTTP method and url endpoint.

## Todos
* Bowerify
* API documentation

## License
Licensed under MIT.


CHANGELOG.md
# 0.1.0

## Features
### pouchConfig
  * pouchConfig is now a provider that allows definition of read and write databases.
	* Added changelog.


src/roo.js
var ngModule = angular.module('vs.ng-roo', ['ng']);

require('./service')(ngModule);

ngModule.provider('rooConfig', function() {
  'use strict';

  var self = this;

  self.$options = {
    auto_compaction: true 
  };

  self.$dwnDbs = [];
  self.$upDbs = [];

  this.$get = function() {
    return {
      getDwnDbs: function() {
        return self.$dwnDbs;
      },
      getUpDbs: function() {
        return self.$upDbs;
      },
      getCouchConfig: function() {
        return self.$couchConfig;
      },
      getOptions: function(){
        return self.$options;
      }
    };
  };

  self.dbs = function(dwnDbs, upDbs) {
    if(dwnDbs && dwnDbs.length > 0) {
      self.$dwnDbs = dwnDbs;
    }

    if(upDbs && upDbs.length > 0) {
      self.$upDbs = upDbs;
    }
  };

  self.couchConfig = function(couchConfig) {
    if(couchConfig) {
      self.$couchConfig = couchConfig;
    }
  };

  self.options = function(options){
    if(options){
      self.$options = _.extend(self.$options, options);
    }
  };

});


Gruntfile.js
var fs = require('fs');

module.exports = function (grunt) {
  'use strict';

  require('load-grunt-tasks')(grunt);

  grunt.initConfig({
    pkg: grunt.file.readJSON('bower.json'),

    language: grunt.option('lang') || 'en',

    meta: {
      banner: '/*!\n * <%= pkg.title || pkg.name %> - v<%= pkg.version %> - ' +
        '<%= grunt.template.today("yyyy-mm-dd") %>\n' +
        ' * <%= pkg.homepage %>\n' +
        ' * Copyright (c) <%= grunt.template.today("yyyy") %> <%= pkg.author.name %>;' +
        ' Licensed <%= _.pluck(pkg.licenses, "type").join(", ") %>\n */\n'
    },

    build_dir: 'dist',

    lib_files: {

      core: [
        'src/roo.js',
        'src/service/LocalStorage.js',
        'src/service/Pouch.js'
      ]
    },

    jshint: {
      options: {
        node: true,
        esnext: false,
        bitwise: true,
        camelcase: false,
        curly: true,
        eqeqeq: true,
        immed: true,
        indent: 2,
        latedef: false,
        noarg: true,
        quotmark: 'single',
        regexp: true,
        undef: true,
        unused: true,
        strict: true,
        trailing: true,
        smarttabs: true,
        newcap: false,
        predef: ['angular', '_', 'PouchDB', 'moment']
      },
      core: {
        files: {
          src: ['<%= lib_files.core %>']
        }
      }
    },

    concat: {
      core: {
        src: ['<%= lib_files.core %>'],
        dest: '<%= build_dir %>/ng-roo.js'
      }
    },

    ngAnnotate: {
      options: {
        singleQuotes: true
      },
      core: {
        src: '<%= concat.core.dest %>',
        dest: '<%= concat.core.dest %>'
      }
    },

    uglify: {
      options: {
        preserveComments: 'some'
      },
      core: {
        files: {
          '<%= build_dir %>/ng-roo.min.js': '<%= concat.core.dest %>'
        }
      }
    },

    version: {
      options: {
        prefix: 'var version\\s+=\\s+[\'"]'
      },
      defaults: {
        src: ['<%= concat.core.dest %>']
      }
    }
  });

	grunt.registerTask('build', [
    'build-all'
  ]);

	grunt.registerTask('build-all', [
    'build:core'
  ]);

  grunt.registerTask('build:core', [
    'jshint:core',
    'concat:core',
    'version',
    'ngAnnotate:core',
    'uglify:core'
  ]);
};


demo/js/todo.js
angular.module('rooTodo', ['vs.ng-roo'])
	.config(function(rooConfigProvider){
		rooConfigProvider.couchConfig({ 'couchUrl': 'http://0.0.0.0:5984' });
	  	rooConfigProvider.dbs(['todos_read'], ['todo_write']);
	})
	.controller('TodoCtrl', function TodoCtrl($scope, $filter, Pouch) {
		'use strict';
		
		var storeOpts = {
			filter: function(){
				return true;
			},
			getParams: function(user){
				return user;
			}
		}

		var store = new Pouch('todos_read');
		store.syncDB('todos_read', {}, storeOpts);
		store.getAll()
		.then(function(docs){
			var todos = $scope.todos = _.pluck(docs.rows, 'doc');
		}).catch(function(){
			console.log(arguments);
		});

		$scope.newTodo = '';
		$scope.editedTodo = null;

		$scope.$watch('todos', function () {
			$scope.remainingCount = $filter('filter')(todos, { completed: false }).length;
			$scope.completedCount = todos.length - $scope.remainingCount;
			$scope.allChecked = !$scope.remainingCount;
		}, true);

		$scope.addTodo = function () {
			var newTodo = {
				title: $scope.newTodo.trim(),
				completed: false
			};

			if (!newTodo.title) {
				return;
			}

			$scope.saving = true;
			// store.insert(newTodo)
			// 	.then(function success() {
			// 		$scope.newTodo = '';
			// 	})
			// 	.finally(function () {
			// 		$scope.saving = false;
			// 	});
		};

		$scope.editTodo = function (todo) {
			$scope.editedTodo = todo;
			// Clone the original todo to restore it on demand.
			$scope.originalTodo = angular.extend({}, todo);
		};

		$scope.saveEdits = function (todo, event) {
			// Blur events are automatically triggered after the form submit event.
			// This does some unfortunate logic handling to prevent saving twice.
			if (event === 'blur' && $scope.saveEvent === 'submit') {
				$scope.saveEvent = null;
				return;
			}

			$scope.saveEvent = event;

			if ($scope.reverted) {
				// Todo edits were reverted-- don't save.
				$scope.reverted = null;
				return;
			}

			todo.title = todo.title.trim();

			if (todo.title === $scope.originalTodo.title) {
				$scope.editedTodo = null;
				return;
			}

			// store[todo.title ? 'put' : 'delete'](todo)
			// 	.then(function success() {}, function error() {
			// 		todo.title = $scope.originalTodo.title;
			// 	})
			// 	.finally(function () {
			// 		$scope.editedTodo = null;
			// 	});
		};

		$scope.revertEdits = function (todo) {
			todos[todos.indexOf(todo)] = $scope.originalTodo;
			$scope.editedTodo = null;
			$scope.originalTodo = null;
			$scope.reverted = true;
		};

		$scope.removeTodo = function (todo) {
			// store.delete(todo);
		};

		$scope.saveTodo = function (todo) {
			// store.put(todo);
		};

		$scope.toggleCompleted = function (todo, completed) {
			if (angular.isDefined(completed)) {
				todo.completed = completed;
			}
			store.put(todo, todos.indexOf(todo))
				.then(function success() {}, function error() {
					todo.completed = !todo.completed;
				});
		};

		$scope.clearCompletedTodos = function () {
			// store.clearCompleted();
		};

		$scope.markAll = function (completed) {
			todos.forEach(function (todo) {
				if (todo.completed !== completed) {
					$scope.toggleCompleted(todo, completed);
				}
			});
		};
	});

src/service/Pouch.js
/*global window: false */
module.exports = function(ngModule) {
  'use strict';

  ngModule.service('Pouch', function ($q, rooConfig, LocalStorageService) {

    function getDB(name) {
      return new PouchDB(name, rooConfig.getOptions());
    }

    /**
     * Function to update the records that are read-only
     * from the server with whatever changes have been
     * made on the client but have not yet been synced
     * to the server, for ui display purposes. This
     * is called currently from the MasterCtrl, so
     * basically on app load. change_id is dbname::id;
     * so, for example, looks something like this:
     *"gemini_cushion_fist::WO-20150226-2949689"
     */
    function shimRecords(downDbName, downDocs) {
      var deferred = $q.defer();
      var promises = [];
      var jorgeNames = rooConfig.getUpDbs();
      // Loop through all of the write databases
      _.each(jorgeNames, function (jorgeName) {
        // Get all the documents in the current write db
        getDB(jorgeName).allDocs({
          include_docs: true
        })
          .then(function (jorgeDocs) {
            if (jorgeDocs.total_rows > 0) {
              _.each(downDocs.rows, function(doc) {
                // The the current write db has any documents, loop through
                // them and see if any of the documents have a change_id that
                // matches the current read-only db
                _.each(jorgeDocs.rows, function (row) {
                  var d = $q.defer();
                  var change_id = row.doc.change_id;
                  var split = change_id.split('::');
                  var dbName = split[0];
                  var id = split[1];
                  if (dbName === downDbName) {
                    // Find if any document matches
                    if(doc.id === id) {
                      _.extend(doc.doc, JSON.parse(row.doc.change));
                    }
                  }
                  d.resolve(doc);
                  promises.push(d.promise);
                });
              });
            }
            $q.all(promises).then(function () {
              deferred.resolve(downDocs);
            });
          });
      });

      return deferred.promise;
    }

    /**
     * Function to update a given read-only record
     * from the server with whatever changes have been
     * made on the client but have not yet been synced
     * to the server, for ui display purposes. This
     * is called currently from the MasterCtrl, so
     * basically on app load. change_id is dbname::id;
     * so, for example, looks something like this:
     *"gemini_cushion_fist::WO-20150226-2949689"
     */
    function shimRecord(downDbName, downDoc) {
      var deferred = $q.defer();
      var jorgeNames = rooConfig.getUpDbs();
      // Loop through all of the write databases
      _.each(jorgeNames, function (jorgeName) {
        // Get all the documents in the current write db
        getDB(jorgeName).allDocs({
          include_docs: true
        }).then(function (jorgeDocs) {
            if (jorgeDocs.total_rows > 0) {
              // The the current write db has any documents, loop through
              // them and see if any of the documents have a change_id that
              // matches the current read-only db
              _.each(jorgeDocs.rows, function (row) {
                var change_id = row.doc.change_id;
                var split = change_id.split('::');
                var dbName = split[0];
                var id = split[1];
                if (dbName === downDbName) {
                  // Find if any document matches
                  if(downDoc._id === id) {
                    _.extend(downDoc, JSON.parse(row.doc.change));
                  }
                }
              });
            }
            deferred.resolve(downDoc);
          });
        });
      return deferred.promise;
    }

  return function (db) {

    this.db = db;

    this.getAll = function (options) {
      var self = this;
      var deferred = $q.defer();

      var options = _.extend({include_docs: true}, options);

      getDB(self.db)
      .allDocs(option)
      .then(function (docs) {
        return shimRecords(self.db, docs);
      }).then(function (docs) {
        deferred.resolve(_.pluck(docs.rows, 'doc'));
      });

      return deferred.promise;
    }

    this.query = function(query, options){
      var self = this;
      var deferred = $q.defer();

      var options = options || {};

      getDB(self.db).query(query, options)
      .then(function (doc) {
        return shimRecord(self.db, doc);
      }).then(function (doc) {
        deferred.resolve(doc);
      });
      return deferred.promise;
    }

    this.getAll = function () {
      var self = this;
      var deferred = $q.defer();
      getDB(self.db).allDocs({
        include_docs: true
      }).then(function (docs) {
        return shimRecords(self.db, docs);
      }).then(function (docs) {
        deferred.resolve(_.pluck(docs.rows, 'doc'));
      });
      return deferred.promise;
    };

    this.get = function (docId) {
      var self = this;
      var deferred = $q.defer();
      getDB(self.db).get(docId, {
        include_docs: true
      }).then(function (doc) {
        return shimRecord(self.db, doc);
      }).then(function (doc) {
        deferred.resolve(doc);
      });
      return deferred.promise;
    };


      /**
       * Retrieve attachment's local url
       * @param {string} docId - Document's id were the attachment exists.
       * @param {string} attrId - The id for the attachment itself.
       */
      this.getAttachment = function (docId, attrId) {
        var self = this;
        var deferred = $q.defer();
        var URL = window.URL;
        getDB(self.db).getAttachment(docId, attrId)
        .then(function (blob) {
          var url = URL.createObjectURL(blob);
          deferred.resolve(url);
        });
        return deferred.promise;
      };

      this.putEntry = function putEntry(table, id, obj, method, endpoint, data, headers, user, successCb, errorCb) {
        var entry = {
            _id: new moment().unix().toString(),
            change_id: '' + table + '::' + id,
            change: JSON.stringify(obj),
            method: method,
            endpoint: endpoint,
            data: data,
            headers: headers,
            user: user
          },
          self = this;

        getDB(self.db).put(entry).then(successCb, errorCb);
      };

      /**
       * Sync db with predefined CouchDB
       * @param {object} user - Any User object to be passed into getParams.
       * @param {object} options - functions to determine params and filters.
       */
      this.syncDB = function syncDB(user, opts) {
        var self = this;
        var deferred = $q.defer();

        // Initialize the remote and local databases
        var remote = new PouchDB(rooConfig.getCouchConfig().couchUrl + '/' + self.db);
        var local = getDB(self.db);
        console.log('syncing', self.db);
        var replicationOptions = {};

        // Set replication options only if they were passed in
        if(opts.filter){
          replicationOptions.filter = opts.filter;
        }
        if(opts.getParams){
          replicationOptions.query_params = opts.getParams(user);
        }

        // Perform replication
        local.replicate.from(remote, replicationOptions)
        .on('complete', function (result) {
          try {
            // Make an entry in the logs
            LocalStorageService.addEntryToLog(user.employeeID, self.db, result);
            deferred.resolve();
          } catch(error) {
            console.log(error);
            deferred.reject(error);
          }
          return deferred.promise;
        });
      };

      this.update = function update(db, obj, id) {
        var deferred = $q.defer();
        var cushiondb = getDB(db);
        cushiondb.get(id).then(function success(doc) {
          var newDoc = _.extend(doc, obj);
          cushiondb.put(newDoc).then(function success(result) {
            deferred.resolve(result);
          }, function error(result) {
            deferred.reject(result);
          });
        }, function error(result) {
          deferred.reject(result);
        });
        return deferred.promise;
      };

      this.sendJorgeDB = function sendJorgeDB(db, user) {
        console.log('syncing ' + db + ' for user ' + user.displayName);
        var deferred = $q.defer();
        var promises = [];
        // Do some stuff here.
        getDB(db).allDocs({
          include_docs: true
        }).then(function (docs) {
          // Make an entry in the logs
          _.forEach(docs.rows, function(row) {
            promises.push(LocalStorageService.addEntryToLog(user.employeeID, db, row));
          });

          $q.all(promises).then(function(result) {
            deferred.resolve(result);
          });
        });
        return deferred.promise;
      };
    };
  });
};


src/service/LocalStorage.js
module.exports = function(ngModule) {
  'use strict';
  ngModule.factory('LocalStorageService', function($window, $q) {

    var lsKey = 'ng-roo';
    var localStorage = $window.localStorage;

    function addEntryToLog(userId, db, data) {
      var ns;
      var newItem;
      var items;
      var deferred = $q.defer();
      ns = getLocalStorageNamespace(userId);
      if(ns) {
        newItem = {
          db: '' + db,
          timestamp: new Date(),
          data: data
        };
        items = JSON.parse(localStorage[ns]);
        items.push(newItem);
        localStorage.setItem(ns, JSON.stringify(items));
        deferred.resolve(items);
      } else {
        deferred.reject('Sorry; HTML5 local storage not supported.');
      }
      return deferred.promise;
    }

    function getLog(userId) {
      var items = [],
        ns = getLocalStorageNamespace(userId),
        deferred = $q.defer();
      if (ns) {
        items = JSON.parse(localStorage[ns]);
        deferred.resolve({ key: ns, items: items });
      } else {
        deferred.reject('Sorry; HTML5 local storage not supported.');
      }
      return deferred.promise;
    }

    function clearLog(userId) {
      var retVal = false,
        ns = getLocalStorageNamespace(userId);
      if(ns) {
        localStorage.removeItem(ns);
        retVal = true;
      }

      return retVal;
    }

    function getLocalStorageNamespace(userId) {
      var retVal;
      var fullKey = 'log.' + lsKey + '.' + userId;

      if (localStorage) {
        if (!localStorage[fullKey]) {
          // create a new localstorage namespace
          localStorage[fullKey] = JSON.stringify([]);
        }
        retVal = fullKey;
      }
      return retVal;
    }

    return {
      addEntryToLog: addEntryToLog,
      getLog: getLog,
      clearLog: clearLog
    };
  });
};


src/PouchDB.js
angular.module('vs.ng-pouch', ['ng']);

angular.module('vs.ng-pouch').provider('pouchConfig', function() {
  var self = this;

  self.$dwnDbs;
  self.$upDbs;

  this.$get = function() {
    return {
      getDwnDbs: function() {
        return self.$dwnDbs;
      },
      getUpDbs: function() {
        return self.$upDbs;
      },
      getCouchConfig: function() {
        return self.$couchConfig;
      }
    }
  };

  self.dbs = function(dwnDbs, upDbs) {
    if(dwnDbs && dwnDbs.length > 0) {
      self.$dwnDbs = dwnDbs;
    }

    if(upDbs && upDbs.length > 0) {
      self.$upDbs = upDbs;
    }
  };

  self.couchConfig = function(couchConfig) {
    if(couchConfig) {
      self.$couchConfig = couchConfig;
    }
  };

});

angular.module('vs.ng-pouch').service('Pouch', function ($q, pouchConfig) {
  'use strict';

  function getDB(name) {
    return new PouchDB(name);
  }

  /**
   * Function to update the records that are read-only
   * from the server with whatever changes have been
   * made on the client but have not yet been synced
   * to the server, for ui display purposes. This
   * is called currently from the MasterCtrl, so
   * basically on app load. change_id is dbname::id;
   * so, for example, looks something like this:
   *"gemini_cushion_fist::WO-20150226-2949689"
   */
  function shimRecords(downDbName, downDocs) {
    var deferred = $q.defer();
    var promises = [];
    var retVal = [];
    var jorgeNames = pouchConfig.getUpDbs();
    // Loop through all of the write databases
    _.each(jorgeNames, function (jorgeName) {
      // Get all the documents in the current write db
      getDB(jorgeName).allDocs({
        include_docs: true
      })
        .then(function (jorgeDocs) {
          if (jorgeDocs.total_rows > 0) {
            _.each(downDocs.rows, function(doc) {
              // The the current write db has any documents, loop through
              // them and see if any of the documents have a change_id that
              // matches the current read-only db
              _.each(jorgeDocs.rows, function (row) {
                var d = $q.defer();
                var change_id = row.doc.change_id;
                var split = change_id.split("::");
                var dbName = split[0];
                var id = split[1];
                if (dbName == downDbName) {
                  // Find if any document matches
                  if(doc.id == id) {
                    _.extend(doc.doc, JSON.parse(row.doc.change));
                  }
                }
                d.resolve(doc);
                promises.push(d.promise);
              });
            });
          }
          $q.all(promises).then(function () {
            deferred.resolve(downDocs);
          });
        });
    });

    return deferred.promise;
  }

  return function (db) {

    this.db = db;

    this.getAll = function () {
      var self = this;
      var deferred = $q.defer();
      getDB(self.db).allDocs({
        include_docs: true
      }).then(function (docs) {
        return shimRecords(self.db, docs);
      }).then(function (docs) {
        deferred.resolve(docs);
      });
      return deferred.promise;
    };

    this.putEntry = function putEntry(table, id, obj, method, endpoint, data, headers, user, successCb, errorCb) {
      var entry = {
          _id: new moment().unix().toString(),
          change_id: '' + table + '::' + id,
          change: JSON.stringify(obj),
          method: method,
          endpoint: endpoint,
          data: data,
          headers: headers,
          user: user
        },
        self = this;

      getDB(self.db).put(entry).then(successCb, errorCb);
    };

    this.syncDB = function syncDB(name, user, opts) {
      var deferred = $q.defer();
      var remote = new PouchDB('http://0.0.0.0:5984' + name);
      //var remote = new PouchDB(couchConfig.host + name);
      var local = new PouchDB(name);
      console.log('syncing', name);
      local.replicate.from(remote, {
        filter: opts.filter,
        query_params: opts.getParams(user)
      }).on('complete', function () {
        defer.resolve();
        return deferred.promise;
      });
    };

    this.update = function update(db, obj, id) {
      var deferred = $q.defer();
      var cushiondb = getDB(db);
      cushiondb.get(id).then(function success(doc) {
        var newDoc = _.extend(doc, obj);
        cushiondb.put(newDoc).then(function success(result) {
          deferred.resolve(result);
        }, function error(result) {
          deferred.reject(result);
        });
      }, function error(result) {
        deferred.reject(result);
      });
      return deferred.promise;
    };

    this.sendJorgeDB = function sendJorgeDB(db, user) {
      console.log('syncing ' + db + ' for user ' + user.displayName);
      var deferred = $q.defer();
      // Do some stuff here.
      deferred.resolve();
      return deferred.promise;
    };
  };
});
